ledger/participant-integration-api/src/main/scala/platform/indexer/ha/HaCoordinator.scala

```diff
@@ -0,0 +1,183 @@
+}
+
+case class HaConfig(
+    mainLockAquireRetryMillis: Long = 500,
+    workerLockAquireRetryMillis: Long = 500,
+    workerLockAquireMaxRetry: Long = 1000,
+    mainLockCheckerPeriodMillis: Long = 1000,
+    indexerLockId: Int = 0x646d6c00, // note 0x646d6c equals ASCII encoded "dml"
+    indexerWorkerLockId: Int = 0x646d6c01,
+    enable: Boolean = false, // TODO ha: remove as stable
+)
+
+object HaCoordinator {
+
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  /** This implementation of the HaCoordinator
+    * - provides a database lock based isolation of the protected executions
+    * - will run the execution at-most once during the entire lifecycle
+    * - will wait infinitely to acquire the lock needed to start the protected execution
+    * - provides a ConnectionInitializer function which is mandatory to execute on all worker connections during execution
+    * - will spawn a polling-daemon to observe continuous presence of the main lock
+    *
+    * @param dataSource to spawn the main connection which keeps the Indexer Main Lock
+    * @param storageBackend is the database-independent abstraction of session/connection level database locking
+    * @param executionContext which is use to execute initialisation, will do blocking/IO work, so dedicated execution context is recommended
+    */
+  def databaseLockBasedHaCoordinator(
+      dataSource: DataSource,
+      storageBackend: DBLockStorageBackend,
+      executionContext: ExecutionContext,
+      scheduler: Scheduler,
+      haConfig: HaConfig,
+  )(implicit loggingContext: LoggingContext): HaCoordinator = {
+    implicit val ec: ExecutionContext = executionContext
+
+    val indexerLockId = storageBackend.lock(haConfig.indexerLockId)
+    val indexerWorkerLockId = storageBackend.lock(haConfig.indexerWorkerLockId)
+    val preemptableSequence = PreemptableSequence(scheduler)
+
+    new HaCoordinator {
+      override def protectedExecution(
+          initializeExecution: ConnectionInitializer => Future[Handle]
+      ): Handle = {
+        def acquireLock(connection: Connection, lockId: LockId, lockMode: LockMode): Lock = {
+          logger.debug(s"Acquiring lock $lockId $lockMode")
+          storageBackend
+            .tryAcquire(lockId, lockMode)(connection)
+            .getOrElse(
+              throw new Exception(s"Cannot acquire lock $lockId in lock-mode $lockMode: lock busy")
+            )
+        }
+
+        def acquireMainLock(connection: Connection): Unit = {
+          acquireLock(connection, indexerLockId, LockMode.Exclusive)
+          ()
+        }
+
+        preemptableSequence.executeSequence { sequenceHelper =>
+          import sequenceHelper._
+          logger.info("Starting databaseLockBasedHaCoordinator")
+          for {
+            mainConnection <- go[Connection](dataSource.getConnection)
+            _ = logger.info("Step 1: creating main-connection - DONE")
```

Comment:
If we hit the kill switch after this line main connection will never be closed

Reply:
I think it might be still ok, because:
* The go wrapper checks the killswitch state before the wrapped future starts. (putting it differently: it does not let the continuation start in case shutdown is ongoing)
* the line 118 is part of the for comprehension, and in this case it will be hit after the go wrapped future already has it's result. At this point we don't care about the kill-switch
* at line 119 the release function is getting registered, but here we also do not care about kill-switches neither, so we just register the release function unaware of an ongoing shutdown process
* at line 124 we start the retry mechanism to grab the indexer main lock. Here we care about the killswitch. Here we will start the shutdown process, but at this point the release function is already registered, so we will release the connection.

With all that said this should be part of unit testing

Reply:
Let's add a TODO then

Reply:
I just added a line to the respective Jira task to capture that testing aspect as well